### PR TITLE
Provide current namespace to InClusterConfig

### LINKED ISSUE #21073: In-container client config doesn't work for many common use cases because no namespace available
This is kind of last minute, but I'd like to consider this for 1.2 because we've had a recent rash of users who want to script this and are effectively blocked.

Today, by default service accounts inject one of their secrets into `/var/run/secrets/kubernetes.io/default`.  That secret enables API access.  We have support in the client config to read that value and provide a default client connection using that service account, which means `kubectl` inside of the container can effectively perform operations as the service account and is easily scriptable.  The one gap today is that `kubectl` does not have access to a default namespace.  Since the point of this injection is to enable easy use of the API by a pod, and most of our APIs that a pod would mutate are namespace scoped, and the service account is scoped to a namespace (although it may have other privileges), the fact that the secret comes with no identifying information as to the namespace means that in order to do "easy" integration with the API, the caller also has to inject the pod namespace via the downward API. 

I would like to add a `namespace` key to each service account token secret, and change in-cluster-config to check for that file and use its value as a default namespace for kubectl and other kubectl like tools.  The secret is a "default injected convention" and the convention is meant to enable easy API integration.  Auto-injecting a downward API volume / env is more complex.  

This enables scenarios like:

* Create a pod that has the command `/bin/sh -c "kubectl rolling-upgrade $SOURCE $DESTINATION && curl -X POST https://myserver/deployment/successful`
* Fetch endpoints for the current service from any pod that has kubectl `kubectl get endpoints db -o jsonpath={.endpoints.subset.addresses}`

### PR DESCRIPTION
**Pull request by liggitt**
Fixes #21073

### CODE REVIEW

File: pkg/client/unversioned/clientcmd/client_config.go [hunk 1]
```diff
@@ -325,12 +327,18 @@ func (inClusterClientConfig) ClientConfig() (*client.Config, error) {
 }
 
 func (inClusterClientConfig) Namespace() (string, error) {
-	// TODO: generic way to figure out what namespace you are running in?
-	// This way assumes you've set the POD_NAMESPACE environment variable
-	// using the downward API.
+	// This way assumes you've set the POD_NAMESPACE environment variable using the downward API.
```
Comment (smarterclayton) [inline pkg/client/unversioned/clientcmd/client_config.go]: Note that this is required for backwards compatibility with old clients

File: pkg/controller/serviceaccount/tokens_controller_test.go [hunk 2]
```diff
@@ -163,6 +165,20 @@ func serviceAccountTokenSecretWithCAData(data []byte) *api.Secret {
...
 
+// serviceAccountTokenSecretWithoutNamespaceData returns an existing ServiceAccountToken secret that lacks namespace data
+func serviceAccountTokenSecretWithoutNamespaceData() *api.Secret {
+	secret := serviceAccountTokenSecret()
+	delete(secret.Data, api.ServiceAccountNamespaceKey)
+	return secret
+}
+
+// serviceAccountTokenSecretWithNamespaceData returns an existing ServiceAccountToken secret with the specified namespace data
+func serviceAccountTokenSecretWithNamespaceData(data []byte) *api.Secret {
```
Comment (smarterclayton) [inline pkg/controller/serviceaccount/tokens_controller_test.go]: Why do you delete the namespace key here, instead of testing?

**Comment by smarterclayton** [comment]
One question on the tests, then needs @kubernetes/kube-api review (since it adds the constant that is part of the secret, and the secret schema is public API), then LGTM

File: pkg/controller/serviceaccount/tokens_controller_test.go (hunk 2 above)
Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller_test.go]: not sure about the question... these are helpers to get test data in the shape I care about... I don't really care whether the key was there or not in `serviceAccountTokenSecretWithoutNamespaceData`

File: pkg/client/unversioned/clientcmd/client_config.go (hunk 1 above)
Comment (liggitt) [inline pkg/client/unversioned/clientcmd/client_config.go]: noted

**Comment by smarterclayton** [comment]
LGTM pending @bgrant0607's review

File: pkg/client/unversioned/clientcmd/client_config.go (hunk 1 above)
Comment (bgrant0607) [inline pkg/client/unversioned/clientcmd/client_config.go]: This should have been prefixed with KUBERNETES_, KUBE_, or somesuch. :-(

Lots of examples use this, but I don't see it actually documented somewhere.

Comment (liggitt) [inline pkg/client/unversioned/clientcmd/client_config.go]: agree :-(

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 3]
```diff
@@ -375,6 +378,10 @@ func (e *TokensController) generateTokenIfNeeded(serviceAccount *api.ServiceAcco
 	if needsCA {
 		secret.Data[api.ServiceAccountRootCAKey] = e.rootCA
 	}
+	// Set the namespace
+	if needsNamespace {
+		secret.Data[api.ServiceAccountNamespaceKey] = []byte(secret.Namespace)
```
Comment (bgrant0607) [inline pkg/controller/serviceaccount/tokens_controller.go]: @pmorie Note that service account secrets are automatically updated.

**Comment by bgrant0607** [comment]
LGTM, but we should document this.

**Comment by lavalamp** [comment]
We need to be sure and build a new kubectl container for 1.2 (for use as a side-car in pods). I think we should be doing that anyway as part of the release.

**Comment by liggitt** [comment]
> I think we should be doing that anyway as part of the release.

I would hope so, if only to get new kubectl features

**Pull request merged**